### DISCUSSION TITLE
XMLExporter - why is the "read_attribute()" used instead of "read_attributes()"?

### DISCUSSION THREAD

**Discussion by xydan83** [FreeOpcUa/opcua-asyncio#699] [Q&A] [answered]
Hi everybody! Analyzing the work of **xmlexporter.py,** I noticed such a thing that instead of reading the attributes in a bunch through the `read_attribute()` method, the attributes are read individually. And in the method of exporting to a file, a loop is simply used, where a list of nodes is passed. Why is `read_attributes()` not used, since one request to the server to get attributes is much faster than multiple requests for one node? Why is such a slow scheme chosen?

![image](https://user-images.githubusercontent.com/86729310/138736052-b77e2eb2-f3c2-4f4c-acba-6a417aa54843.png)

**Comment by oroulet** [comment, marked as answer]
Historical reasons. The XML exporter at some point was very simple code . Now it is a monster.
PR welcome byit be careful not to make things too complicated.
